### ISSUE TITLE
App behind ALB Fails SDK Access

### ISSUE THREAD

**Issue by forbode** [boxfuse/boxfuse-issues#188]
Up until recently I have been using the ELB. The switch to ALB has been a learning curve, but I have hit a new stumbling block. 
The problem is that the app behind the ALB does not to push to Cloudwatch Logs and AWS API access is timing out. 

The AWS Profile I am using is the same as when I was using the ELB and it worked well. (I could access the AWS API Services such as S3). In this respect I know that the Policy is correct so I've ruled out a policy issue. 

The Security Group Inbound and outbound is also correct (outbound is wide open). 

I'm wondering if there is simply something I don't know about the ALB or the recent Boxfuse Maven upgrade has broken my deploy. 
Ideas?

**Comment by forbode** [comment]
Out of nowhere a few 5 log entries showed up. I'm not sure why or how. 

![screen shot 2017-10-30 at 8 16 24 pm](https://user-images.githubusercontent.com/5214966/32205909-47ca0fc4-bdaf-11e7-8d7a-7d67843984ef.png)

**Comment by axelfontaine** [comment]
Is this still an issue or is it resolved? If it is, can you share your exact configuration?

Note that auto-scaling group creation is identical for ELB and ALB apps, so there should be no difference there in terms of instance profile handling.

**Comment by forbode** [comment]
At this moment I cannot get the exact configuration I can in the early morning.  When you say you exact are you looking for exact Security group ID’s, subnet, etc?


It is still an issue.

**Comment by axelfontaine** [comment]
Exact IDs don't matter. More which exact parameters have you specified from the command-line and in your boyfuse.conf file + what have you configured in the Console for that environment.

**Comment by forbode** [comment]
I'm using Maven. 

- Custom environment (boxfuse.env)
- Capacity 1:t2.medium
- Target Group
- EnvVars.SPRING_PROFILES_ACTIVE
- Security Group
- Subnets (2 to be exact. 2 are required for ALB)
- App.Type is load-balanced-https
- Tls.type is acm
- Instance.profile
- log.type is cloudwatch-logs

I checked the Cloudwatch Logs again and nothing new has showed.

**Comment by forbode** [comment]
It seems that the same issue is happening with Worker applications that auto scale as well. The two applications that actually get logs into Cloudwatch (using the Boxfuse logs agent setup) are single instance assigned to a specific elastic IP. 

I'll be digging in trying to find out what I can.

**Comment by boxfuse** [comment]
Is your instance profile correctly associated with the instances?

**Comment by forbode** [comment]
Yes, the instances behind the ALB are associated properly  to the IAM Role. They are using the same role as the EC2 instances that work (which are single instance)

**Comment by forbode** [comment]
Cloudwatch logs agent times out. I suspect that if it didn't get credentials it would throw a different error. The timeout seems to indicate a networking issue? 

![screen shot 2017-10-31 at 9 11 22 am](https://user-images.githubusercontent.com/5214966/32235031-99b34870-be1b-11e7-95f4-39f094be20b0.png)

**Comment by forbode** [comment]
I removed the EC2 instance from the target group (associated with the ELB) and assigned it an Elastic IP and suddenly... Logs files come through. In addition, the posts that use the AWS SDK instantly return. 

The point is there is something going on when the application is EITHER behind an ALB or not associated with an Elastic IP. 

I will add it back to the target group and leave the EIP and see what happens.

**Comment by forbode** [comment]
After adding the EC2 instance back into the ALB (but keeping the EIP) it continued to work just fine. The logs agents works fast along with SDK access not timing out. 

So it's down to the EC2 instance REQUIRING a public IP address if it needs AWS SDK access. This makes no sense whatsoever because the purpose of the ALB is to be the public facing side of many servers. 

Something unusual is happening with the networking that causes this to happen.

**Comment by forbode** [comment]
I am now nearly certain that is has to do with this:
https://aws.amazon.com/premiumsupport/knowledge-center/public-load-balancer-private-ec2/

**Comment by axelfontaine** [comment]
Thanks for the investigation. This probably makes sense as the SDK needs to be able to communicate with the AWS service and those network packets must be routed somehow.
